File: backend/app/schemas/team.py
import uuid
from datetime import datetime

from pydantic import AliasChoices, BaseModel, Field

from app.models.store import MemberRole, InviteStatus


class TeamInviteCreate(BaseModel):
    invited_email: str = Field(
        min_length=3,
        max_length=255,
        validation_alias=AliasChoices("invited_email", "invitedEmail", "email")
    )
    role: MemberRole = MemberRole.admin


class TeamInviteResponse(BaseModel):
    id: uuid.UUID
    store_id: uuid.UUID
    invited_by_user_id: uuid.UUID
    invited_email: str
    role: MemberRole
    status: InviteStatus
    token: str
    invite_link: str = ""
    expires_at: datetime
    accepted_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TeamMemberResponse(BaseModel):
    id: uuid.UUID
    store_id: uuid.UUID
    user_id: uuid.UUID
    name: str | None = None
    email: str | None = None
    role: MemberRole
    role_name: str
    role_priority: int
    permissions: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TeamInviteAcceptResponse(BaseModel):
    store_id: uuid.UUID
    role: MemberRole
    status: str


class TeamRoleCreate(BaseModel):
    name: str = Field(min_length=2, max_length=64)
    description: str | None = Field(default=None, max_length=500)
    priority: int = Field(ge=0, le=100)
    permissions: list[str] = Field(default_factory=list)


class TeamRoleUpdate(BaseModel):
    description: str | None = Field(default=None, max_length=500)
    priority: int | None = Field(default=None, ge=0, le=100)
    permissions: list[str] | None = None


class TeamRoleResponse(BaseModel):
    id: uuid.UUID
    store_id: uuid.UUID
    name: str
    description: str | None = None
    priority: int
    permissions: list[str] = Field(default_factory=list)
    is_system: bool
    is_editable: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TeamMemberRoleUpdate(BaseModel):
    role_id: uuid.UUID


class TeamMemberRoleBulkItem(BaseModel):
    member_id: uuid.UUID
    role_id: uuid.UUID


class TeamMemberRoleBulkApplyRequest(BaseModel):
    updates: list[TeamMemberRoleBulkItem] = Field(default_factory=list)


class TeamMemberRoleBulkApplyResponse(BaseModel):
    applied_count: int

File: frontend/queries/team.ts
import { fetchWithAccessToken } from "@/lib/auth-fetch";
import { denormalizeRequest } from "@/lib/normalize-response";
import { queryOptions, useQuery } from "@tanstack/react-query";

export type TeamMember = {
    id: string;
    storeId: string;
    userId: string;
    name: string | null;
    email: string | null;
    role: "owner" | "admin" | "staff";
    roleName: string;
    rolePriority: number;
    permissions: string[];
    createdAt: string;
    updatedAt: string;
};

export type TeamRole = {
    id: string;
    storeId: string;
    name: string;
    description: string | null;
    priority: number;
    permissions: string[];
    isSystem: boolean;
    isEditable: boolean;
    createdAt: string;
    updatedAt: string;
};

export type TeamInvite = {
    id: string;
    storeId: string;
    invitedByUserId: string;
    invitedEmail: string;
    role: "owner" | "admin" | "staff";
    status: "pending" | "accepted" | "revoked" | "expired";
    token: string;
    inviteLink: string;
    expiresAt: string;
    acceptedAt: string | null;
    createdAt: string;
    updatedAt: string;
};

export type InvitePayload = {
    invitedEmail: string;
    role: "admin" | "staff";
};

export type CreateRolePayload = {
    name: string;
    description?: string;
    priority: number;
    permissions: string[];
};

export type UpdateRolePayload = {
    description?: string;
    priority?: number;
    permissions?: string[];
};

export type ApplyMemberRolesPayload = {
    updates: { memberId: string; roleId: string }[];
};

export const teamMembersQueryOptions = (storeId: string | undefined) =>
    queryOptions({
        queryKey: ["team-members", storeId],
        queryFn: () => fetchWithAccessToken<TeamMember[]>(`/stores/${storeId}/members`),
        enabled: !!storeId,
    });

export const teamInvitesQueryOptions = (storeId: string | undefined) =>
    queryOptions({
        queryKey: ["team-invites", storeId],
        queryFn: () => fetchWithAccessToken<TeamInvite[]>(`/stores/${storeId}/invites`),
        enabled: !!storeId,
    });

export const teamRolesQueryOptions = (storeId: string | undefined) =>
    queryOptions({
        queryKey: ["team-roles", storeId],
        queryFn: () => fetchWithAccessToken<TeamRole[]>(`/stores/${storeId}/roles`),
        enabled: !!storeId,
    });

export function useTeamMembersQuery(storeId: string | undefined) {
    return useQuery(teamMembersQueryOptions(storeId));
}

export function useTeamInvitesQuery(storeId: string | undefined) {
    return useQuery(teamInvitesQueryOptions(storeId));
}

export function useTeamRolesQuery(storeId: string | undefined) {
    return useQuery(teamRolesQueryOptions(storeId));
}

export async function createInvite(storeId: string | undefined, payload: InvitePayload) {
    return fetchWithAccessToken<TeamInvite>(`/stores/${storeId}/invites`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(denormalizeRequest(payload)),
    });
}

export async function revokeInvite(storeId: string | undefined, inviteId: string) {
    return fetchWithAccessToken<void>(`/stores/${storeId}/invites/${inviteId}`, {
        method: "DELETE",
    });
}

export async function createRole(storeId: string | undefined, payload: CreateRolePayload) {
    return fetchWithAccessToken<TeamRole>(`/stores/${storeId}/roles`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(denormalizeRequest(payload)),
    });
}

export async function updateRole(storeId: string | undefined, roleId: string, payload: UpdateRolePayload) {
    return fetchWithAccessToken<TeamRole>(`/stores/${storeId}/roles/${roleId}`, {
        method: "PATCH",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(denormalizeRequest(payload)),
    });
}

export async function deleteRole(storeId: string | undefined, roleId: string) {
    return fetchWithAccessToken<void>(`/stores/${storeId}/roles/${roleId}`, {
        method: "DELETE",
    });
}

export async function applyMemberRoles(storeId: string | undefined, payload: ApplyMemberRolesPayload) {
    return fetchWithAccessToken<{ appliedCount: number }>(`/stores/${storeId}/members/roles/apply`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(denormalizeRequest(payload)),
    });
}

export async function acceptInvite(token: string) {
    return fetchWithAccessToken<{ storeId: string; role: "owner" | "admin" | "staff"; status: string }>(
        `/team/invites/accept/${token}`,
        {
            method: "POST",
        }
    );
}
